fineract-investor/src/main/java/org/apache/fineract/investor/service/LoanAccountOwnerTransferServiceImpl.java

```diff
@@ -0,0 +1,210 @@
+
+        businessEventNotifierService.notifyPostBusinessEvent(new LoanOwnershipTransferBusinessEvent(buybackTransfer, loan));
+        businessEventNotifierService.notifyPostBusinessEvent(new LoanAccountSnapshotBusinessEvent(loan));
+    }
+
+    private ExternalAssetOwnerTransfer createCancelledTransfer(ExternalAssetOwnerTransfer pendingTransfer, LocalDate settlementDate,
+            ExternalTransferSubStatus subStatus) {
+        ExternalAssetOwnerTransfer cancelledTransfer = new ExternalAssetOwnerTransfer();
+        cancelledTransfer.setOwner(pendingTransfer.getOwner());
+        cancelledTransfer.setExternalId(pendingTransfer.getExternalId());
+        cancelledTransfer.setStatus(CANCELLED);
+        cancelledTransfer.setSubStatus(subStatus);
+        cancelledTransfer.setSettlementDate(settlementDate);
+        cancelledTransfer.setLoanId(pendingTransfer.getLoanId());
+        cancelledTransfer.setExternalLoanId(pendingTransfer.getExternalLoanId());
+        cancelledTransfer.setPurchasePriceRatio(pendingTransfer.getPurchasePriceRatio());
+        cancelledTransfer.setEffectiveDateFrom(settlementDate);
+        cancelledTransfer.setEffectiveDateTo(settlementDate);
+        return externalAssetOwnerTransferRepository.save(cancelledTransfer);
+    }
+
+    private ExternalAssetOwnerTransfer createDeclinedTransfer(ExternalAssetOwnerTransfer pendingSaleTransfer, LocalDate settlementDate,
+            Loan loan) {
+        ExternalAssetOwnerTransfer declinedTransfer = new ExternalAssetOwnerTransfer();
+        declinedTransfer.setOwner(pendingSaleTransfer.getOwner());
+        declinedTransfer.setExternalId(pendingSaleTransfer.getExternalId());
+        declinedTransfer.setStatus(DECLINED);
+        declinedTransfer.setSubStatus(isBiggerThanZero(loan.getTotalOverpaid()) ? BALANCE_NEGATIVE : BALANCE_ZERO);
+        declinedTransfer.setSettlementDate(settlementDate);
+        declinedTransfer.setLoanId(pendingSaleTransfer.getLoanId());
+        declinedTransfer.setExternalLoanId(pendingSaleTransfer.getExternalLoanId());
+        declinedTransfer.setPurchasePriceRatio(pendingSaleTransfer.getPurchasePriceRatio());
+        declinedTransfer.setEffectiveDateFrom(settlementDate);
+        declinedTransfer.setEffectiveDateTo(settlementDate);
+        return externalAssetOwnerTransferRepository.save(declinedTransfer);
+    }
+
+    private void updatePendingTransfer(ExternalAssetOwnerTransfer pendingTransfer, LocalDate settlementDate) {
+        pendingTransfer.setEffectiveDateTo(settlementDate);
+        externalAssetOwnerTransferRepository.save(pendingTransfer);
+    }
+
+    private ExternalAssetOwnerTransfer updatePendingBuybackTransfer(Loan loan, ExternalAssetOwnerTransfer buybackTransfer,
+            LocalDate settlementDate) {
+        buybackTransfer.setEffectiveDateTo(settlementDate);
+        buybackTransfer.setSubStatus(isBiggerThanZero(loan.getTotalOverpaid()) ? BALANCE_NEGATIVE : BALANCE_ZERO);
```

Comment:
i was thinking and we dont need the substatus.

Reply:
Okay. Only for this case?

We have the other case: when the loan is closed, and we have a buyback and a sale transfer (with different settlement date). Those transfers will be closed, and a substatus will be either BALANCE_NEGATIVE or BALANCE_ZERO. Do we need the substatus there?

Reply:
when it is CANCELLED / DECLINED, yes we need the substatus as clarification.

Reply:
Thanks, everything is clear now :) !